exonum-java-binding-qa-service/src/test/java/com/exonum/binding/qaservice/ApiControllerIntegrationTest.java

```diff
@@ -427,6 +473,256 @@ void getBlockTransactions(VertxTestContext context) {
+
+    get(BLOCKCHAIN_TRANSACTION_LOCATION_PATH.replace(":" + MESSAGE_HASH_PARAM, HASH_STRING))
+        .send(context.succeeding(response -> context.verify(() -> {
+          assertThat(response.statusCode())
+              .isEqualTo(HTTP_OK);
+
+          String body = response.bodyAsString();
+          Object actualLocation = JSON_SERIALIZER
+              .fromJson(body, new TypeToken<TransactionLocation>() {
+              }.getType());
+          assertThat(actualLocation).isEqualTo(transactionLocation);
+
+          context.completeNow();
+        })));
+  }
+
+  @Test
+  void getNonexistentTxLocation(VertxTestContext context) {
+    HashCode messageHash = HashCode.fromString(HASH_STRING);
+    when(qaService.getTxLocation(messageHash)).thenReturn(Optional.empty());
+
+    get(BLOCKCHAIN_TRANSACTION_LOCATION_PATH.replace(":" + MESSAGE_HASH_PARAM, HASH_STRING))
+        .send(context.succeeding(response -> context.verify(() -> {
+          assertThat(response.statusCode())
+              .isEqualTo(HTTP_NOT_FOUND);
+
+          context.completeNow();
+        })));
+  }
+
+  @Test
+  void getBlocks(VertxTestContext context) {
+    Map<HashCode, Block> blocks = Maps.uniqueIndex(
+        Arrays.asList(createBlock(1L), createBlock(2L)),
+        Block::getBlockHash);
+
+    when(qaService.getBlocks()).thenReturn(blocks);
+
+    get(BLOCKCHAIN_BLOCKS_PATH)
+        .send(context.succeeding(response -> context.verify(() -> {
+          assertThat(response.statusCode())
+              .isEqualTo(HTTP_OK);
+
+          String body = response.bodyAsString();
+          Object actualBlocks = JSON_SERIALIZER
+              .fromJson(body, new TypeToken<Map<HashCode, Block>>() {
+              }.getType());
+          assertThat(actualBlocks).isEqualTo(blocks);
+
+          context.completeNow();
+        })));
+  }
+
+  @Test
+  void getBlock(VertxTestContext context) {
+    Block block = createBlock(1L);
+
+    long blockHeight = block.getHeight();
+    when(qaService.getBlock(blockHeight)).thenReturn(block);
+
+    get(BLOCKCHAIN_BLOCK_PATH.replace(":" + BLOCK_ID_PARAM, Long.toString(blockHeight)))
+        .send(context.succeeding(response -> context.verify(() -> {
+          assertThat(response.statusCode())
+              .isEqualTo(HTTP_OK);
+
+          String body = response.bodyAsString();
+          Object actualBlock = JSON_SERIALIZER
+              .fromJson(body, new TypeToken<Block>() {
+              }.getType());
+          assertThat(actualBlock).isEqualTo(block);
+
+          context.completeNow();
+        })));
+  }
+
+  @Test
+  void getNonexistentBlock(VertxTestContext context) {
+    long blockHeight = 1L;
+
+    when(qaService.getBlock(blockHeight)).thenThrow(new IllegalArgumentException());
+
+    get(BLOCKCHAIN_BLOCK_PATH.replace(":" + BLOCK_ID_PARAM, HASH_STRING))
+        .send(context.succeeding(response -> context.verify(() -> {
+          assertThat(response.statusCode())
+              .isEqualTo(HTTP_NOT_FOUND);
```

Comment:
Shan't it be bad request if an exception occurs?

Reply:
Yes, that's what I wanted to change, as we moved from Optional to throwing exceptions. Although it still kind of makes sense to return `HTTP_NOT_FOUND` in case of `IndexOutOfBounds`, no?